### PR TITLE
#23605: Add docker image ref to downloading image task event

### PR DESCRIPTION
**Pull request by tskorupka**

### CODE REVIEW

**Comment by tskorupka** [comment]
Can someone point me out where I should add additional tests for those Task Events?

File: drivers/docker/driver.go [hunk 1]
```diff
@@ -653,7 +653,7 @@ func (d *Driver) pullImage(task *drivers.TaskConfig, driverConfig *TaskConfig, r
 		AllocID:   task.AllocID,
 		TaskName:  task.Name,
 		Timestamp: time.Now(),
-		Message:   "Downloading image",
+		Message:   fmt.Sprintf("Downloading image %s", dockerImageRef(repo, tag)),
 		Annotations: map[string]string{
 			"image": dockerImageRef(repo, tag),
 		},
```
Comment (tgross) [inline drivers/docker/driver.go:659]: This makes me realize that we already have all the data we need here in the annotations and it already shows up in the API:

```
$ nomad operator api "/v1/allocation/018da10f-16c2-7bba-7905-e0bea88aebc7" | jq '.TaskStates["task"].Events[2]'
{
  "Details": {
    "image": "busybox:1"
  },
  "DiskLimit": 0,
  "DisplayMessage": "Downloading image",
  "DownloadError": "",
  "DriverError": "",
  "DriverMessage": "Downloading image",
  "ExitCode": 0,
  "FailedSibling": "",
  "FailsTask": false,
  "GenericSource": "",
  "KillError": "",
  "KillReason": "",
  "KillTimeout": 0,
  "Message": "",
  "RestartReason": "",
  "SetupError": "",
  "Signal": 0,
  "StartDelay": 0,
  "TaskSignal": "",
  "TaskSignalReason": "",
  "Time": 1722525403089176300,
  "Type": "Driver",
  "ValidationError": "",
  "VaultError": ""
}
```

We're just not displaying these details in the UI or CLI. That seems like a better place to do this rather than repeating state that already exists, doesn't it?

Comment (tskorupka) [reply drivers/docker/driver.go:659]: Yes it makes sense, in fact this DriverMessage/DisplayMessage can be used for localisation, hence displaying it only on the UI side is better approach, let me change that.

**Comment by tskorupka** [comment]
Hey @tgross it took me some time,

TL;DR;

Should I keep working on this, or should I change the way how this will be represented?

I did not want to implement on my own any custom user interface, so I did check how tab for deployment of job looks like, and tried to recreate similar experience from there.

Happy to change it to whatever you approve.

This PR would still require adding e2e tests and providing fixtures.

Also I found that in many places, annotations of task event, so details as they are being returned from API and how the field is named in frontend - are having sometimes duplicates from message.

Should I refactor task events and somehow remove those annotations somehow or rather the message should be refactored to shorter version, so more information can be found by expanding the details?

Also I was thinking that possibly it would be better if we can move deployment history as a separate section in job?

Screenshot 2024-08-02 at 23 21 35

Also I found another view of Recent Events for a Task, should we align it there too? (Show details button added etc.)

Screenshot 2024-08-02 at 23 24 06

The outcome is as in screenshots attached.

Screenshot 2024-08-02 at 23 12 05
Screenshot 2024-08-02 at 23 12 13

**Comment by tgross** [comment]
Thanks for this @tskorupka! I'm going to tag-in @philrenaud on reviewing this, as he's our web UI expert and he'll definitely have more useful thoughts on the UI changes than I will. :grin:

File: ui/app/models/task-event.js [hunk 2]
```diff
@@ -17,11 +20,16 @@ export default class TaskEvent extends Fragment {
   @attr('date') time;
   @attr('number') timeNanos;
   @attr('string') displayMessage;
+  @attr({ defaultValue: () => ({}) }) details;
 
   get message() {
     let message = simplifyTimeMessage(this.displayMessage);
     return message;
```
Comment (philrenaud) [inline ui/app/models/task-event.js:27]: Consider appending the image name via conditional check here

**Comment by tskorupka** [comment]
Hey @philrenaud, thanks for such comprehensive answer.

I agree with your suggestions, but I also think upfront and about possible future requests of having such details being shown, that is why I was thinking about having a generic way of displaying that kind of information.

There could be a lot of other/different information being stored in annotations/details, that users could look for, and we could leverage that functionality to refactor task event messages to show less content and maybe restrict it to specific amount of characters, so it would not be flooded with information there.

Adding specific methods to cover edge cases where the annotation/detail should be displayed will only introduce chaos to the code, that is something that I wanted to skip, and it could grow bigger.

If searching is one of the cases where this edge case should be covered by appending the detail into display message field, I would rather suggest extending search method with lookup by detail.

Just let me know if this is meaningful, if not I will rewrite the code by the requested by you changes.

Kind Regards,
Tomasz

**Comment by tskorupka** [comment]
Hey @philrenaud, done.

**Comment by philrenaud** [comment]
Change looks great, thanks!

There's an acceptance test for [searching deployment history](https://github.com/hashicorp/nomad/blob/eacf47a48472ec22b95aee897ab446d7bec6e8a3/ui/tests/acceptance/job-status-panel-test.js#L750-L809) and I'd consider adding a test there. We use Mirage for mocking data in our acceptance tests, and the one for task-events is at https://github.com/hashicorp/nomad/blob/859087640a74fd074f4a86ba2ddef71e46ae4017/ui/mirage/factories/task-event.js#L21-L22

This looks like it might be a little tricky and esoteric to Ember/Mirage, so please let me know if you'd like help with this test.

Finally, we have a convention of adding changelog entries for contributions; would you please run `make cl` for this PR before final submission?

Thank you so much for all of this, it's a really nice change!

**Comment by tskorupka** [comment]
Hey @philrenaud,

Thanks, great news.

> There's an acceptance test for [searching deployment history](https://github.com/hashicorp/nomad/blob/eacf47a48472ec22b95aee897ab446d7bec6e8a3/ui/tests/acceptance/job-status-panel-test.js#L750-L809) and I'd consider adding a test there. We use Mirage for mocking data in our acceptance tests, and the one for task-events is at https://github.com/hashicorp/nomad/blob/859087640a74fd074f4a86ba2ddef71e46ae4017/ui/mirage/factories/task-event.js#L21-L22

Adding assertion and extending existing test is fine?

> Finally, we have a convention of adding changelog entries for contributions; would you please run `make cl` for this PR before final submission?

Will do.

Kind Regards,
Tomasz

**Comment by philrenaud** [comment]
> Adding assertion and extending existing test is fine?

Yep, that's exactly what I'd do here.

**Comment by tskorupka** [comment]
Still relevant, worthwhile adding?

**Comment by philrenaud** [comment]
Hi @tskorupka, I think still perfectly relevant. If you'd like, I can take a shot at adding the acceptance test assertion and changelog entry. Would you mind if I did?

**Comment by tskorupka** [comment]
Hey @philrenaud, it would be great if you do, sorry, I was offline for a while, I definitely look forward to adding this myself, but also as a collaborative effort I would be thankful ^^.

**Comment by shantanugadgil** [comment]
possibly related to https://github.com/hashicorp/nomad/issues/11024 and https://github.com/hashicorp/nomad/issues/11009
